### PR TITLE
add metric-alarm evaluation to Cloudwatch

### PR DESCRIPTION
**Pull request by steffyP**
This PR refers to https://github.com/localstack/localstack/issues/5446 and implements basic cloudwatch metric-alarm evaluation:

supports metric-alarm with the following characteristics:
- "Statistic": Maximum, SampleCount, Sum, Minimum, Average
- "ComparisonOperator": GreaterThanOrEqualToThreshold, GreaterThanThreshold, LessThanThreshold, LessThanOrEqualToThreshold
- "[Alarm|OK|InsufficentData]Actions": SNS Topic

Note: in order to correctly filter for Dimensions, [this PR in Moto](https://github.com/spulec/moto/pull/5041) is required.

### CODE REVIEW

File: tests/unit/test_cloudwatch.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            ):
+                calculate_alarm_state("helloworld")
+                assert len(mock_client.mock_calls) == expected_calls
+                if expected_calls != 0:
+                    expected_calls = [
+                        call.set_alarm_state(
+                            AlarmName="test-alarm", StateValue=expected_state, StateReason=ANY
+                        )
+                    ]
+                    mock_client.assert_has_calls(expected_calls)
```
Comment (thrau) [inline tests/unit/test_cloudwatch.py:228]: this test could be cleaned up by using the `monkeypatch` fixture that [pytest provides](https://docs.pytest.org/en/6.2.x/monkeypatch.html#simple-example-monkeypatching-functions)

**Review by thrau — APPROVED** [review]
fantastic PR! super clean code, very readable and thorough testing. pretty cool testing strategy as well! :100: overall

just a couple of minor nitpicks.

**Comment by thrau** [comment]
waiting for @silv-io's review and then let's merge! :shipit:

File: localstack/services/cloudwatch/provider.py [hunk 2]
```diff
@@ -226,19 +254,49 @@ def put_metric_alarm(
...
+
+        # missing will be the default, when not set (but it will not explicitly be set)
+        if not request.get("TreatMissingData", "missing") in [
+            "breaching",
+            "notBreaching",
+            "ignore",
+            "missing",
+        ]:
+            raise ValidationError(
+                f"The value {request['TreatMissingData']} is not supported for TreatMissingData parameter. Supported values are [breaching, notBreaching, ignore, missing]."
```
Comment (silv-io) [inline localstack/services/cloudwatch/provider.py:266]: **nit:** Hard-coding the supported values in this error message might lead to typos further down the line if those values are changed.

File: tests/unit/test_cloudwatch.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        X = breaking threshold
+        - = no data available
+        |Test   | Data points  | MISSING          | IGNORE      | BREACHING | NOT BREACHING | # missing datapoints |
+        |-------|--------------|------------------|-------------|-----------|---------------| ---------------------|
+        |TEST_1 | 0 - X - X    | OK               | OK          | OK        | OK            | 0                    |
+        |TEST_2 | 0 - - - -    | OK               | OK          | OK        | OK            | 2                    |
+        |TEST_3 | - - - - -    | INSUFFICIENT_DATA| Retain state| ALARM     | OK            | 3                    |
+        |TEST_4 | 0 X X - X    | ALARM            | ALARM       | ALARM     | ALARM         | 0                    |
+        |TEST_5 | - - X - -    | ALARM            | Retain state| ALARM     | OK            | 2 -> Premature alarm |
+        """
```
Comment (silv-io) [inline tests/unit/test_cloudwatch.py:89]: beautiful

**Review by silv-io — APPROVED** [review]
LGTM! :)

File: tests/unit/test_cloudwatch.py (hunk 1 above)
Comment (thrau) [reply tests/unit/test_cloudwatch.py]: i cleaned it up with our own patch utility :-)

File: localstack/services/cloudwatch/alarm_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+            )
+            return
+
+        period = alarm_details["Period"]
+        evaluation_periods = alarm_details["EvaluationPeriods"]
+        schedule_period = evaluation_periods * period
+
+        def on_error(e):
+            LOG.error(f"Error executing scheduled alarm: {e}")
+            LOG.error(traceback.format_exc())
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/services/cloudwatch/alarm_scheduler.py]: committed the change

**Pull request merged**